Added definitions for oneDAL namespaces (#144)
oneapi-src/oneAPI-spec: source/elements/oneDAL/source/common_interface/header_files_and_namespaces.rst

``onedal``               The namespace of the library that contains externally exposable data types, processing and service functionality of oneDAL. 
``onedal::%ALGORITHM%``  The namespace of the algorithm. All classes and structures related to that algorithm shall be defined within a particular namespace. To define a specific algorithm, the string %ALGORITHM% should be substituted with its name, for example, ``onedal::kmeans`` or ``onedal::knn``.
``onedal::misc``         The namespace that contains miscellaneous data types and functionality intended to be used by oneDAL algorithms and applications for algorithm customization and optimization on various stages of the analytical pipeline.
``%PARENT%::detail``     The namespace that contains implementation details of the data types and functionality for the parent namespace. The namespace can be on any level in the namespace hierarchy. To define a specific namespace, the string %PARENT% should be substituted with the namespace for which the details are provided, for example, ``onedal::detail`` or ``onedal::kmeans::detail``. The application shall not use any data types nor call any functionality located in the ``detail`` namespaces.

> =======================  =======================================================================================================

Reviewer: this should work 

```
.. list-table:: 
   :widths:  10 50
   :header-rows: 1

   * - namespace
     - oneDAL content
   * - ``onedal``
     - The namespace of the library that contains externally exposable data types, processing and service functionality of oneDAL. 
   * - ``onedal::%ALGORITHM%``
     - The namespace of the algorithm. 
       All classes and structures related to that algorithm shall be defined within a particular namespace.
       
       To define a specific algorithm, the string %ALGORITHM% should be substituted with its name, 
       for example, ``onedal::kmeans`` or ``onedal::knn``.

   * - ``onedal::misc``
     - The namespace that contains miscellaneous data types and functionality intended to be used by oneDAL algorithms
       and applications for algorithm customization and optimization on various stages of the analytical pipeline.
   * - ``%PARENT%::detail`` 
     - The namespace that contains implementation details of the data types and functionality for the parent namespace.
       The namespace can be on any level in the namespace hierarchy.
       
       To define a specific namespace, the string %PARENT% should be substituted with the namespace for which the details are provided,
       for example, ``onedal::detail`` or ``onedal::kmeans::detail``. 
       
       The application shall not use any data types nor call any functionality located in the ``detail`` namespaces.
```

============
Header files
============

==========
Namespaces
==========

The functionality of oneDAL library is represented with a system of the C++ namespaces described below:

=======================  =======================================================================================================
namespace                oneDAL content
=======================  =======================================================================================================
``onedal``               The namespace of the library that contains externally exposable data types, processing and service functionality of oneDAL. 

> ``onedal::%ALGORITHM%``  The namespace of the algorithm. All classes and structures related to that algorithm shall be defined within the namespace. The string %ALGORITHM% shall be substituted to define specific algorithm, for example, ``onedal::kmeans``, ``onedal::knn``. 

Reviewer 2: the same comment about are/shall/should. Indeed, this is a req to an implementer to have classes/functions in the respective namespace
For string substitution - it is about the correct interpretation of the description, not about a req, thus, should works

> ``onedal::%ALGORITHM%``  The namespace of the algorithm. All classes and structures related to that algorithm shall be defined within the namespace. The string %ALGORITHM% shall be substituted to define specific algorithm, for example, ``onedal::kmeans``, ``onedal::knn``. 

Author: "this namespace" is a bit confusing for me because there are many algorithm namespaces - not a single one.

> ``onedal::%ALGORITHM%``  The namespace of the algorithm. All classes and structures related to that algorithm shall be defined within the namespace. The string %ALGORITHM% shall be substituted to define specific algorithm, for example, ``onedal::kmeans``, ``onedal::knn``. 

Reviewer: `this namespace` is the same as `the namespace`, but clearer for the reader

> ``onedal::%ALGORITHM%``  The namespace of the algorithm. All classes and structures related to that algorithm shall be defined within the namespace. The string %ALGORITHM% shall be substituted to define specific algorithm, for example, ``onedal::kmeans``, ``onedal::knn``. 

Reviewer: change `The namespace of the algorithm` -> `The namespace of a particular algorithm`

this gives you "not a single one" notion

``onedal::misc``         The namespace that contains miscellaneous data types and functionality which are intended for the use by the oneDAL algorithms or by the applications for customization of the algorithms or optimization of the stages of analytical pipeline.

> ``%PARENT%::detail``     The namespace that contains implementation details of the data types and functionality for parent namespace. The namespace can be on any level of namespace hierarchy. The string %PARENT% shall be substituted to define specific namespace, for example, ``onedal::detail``, ``onedal::kmeans::detail``. The application shall not use any data types nor call any functionality located in ``detail`` namespaces.

Reviewer 2: Re applications: it's the requirement for the calling app which serves as the note/indicator to the spec implementer

> ``%PARENT%::detail``     The namespace that contains implementation details of the data types and functionality for parent namespace. The namespace can be on any level of namespace hierarchy. The string %PARENT% shall be substituted to define specific namespace, for example, ``onedal::detail``, ``onedal::kmeans::detail``. The application shall not use any data types nor call any functionality located in ``detail`` namespaces.

Author: "Application should not" doesn't look as an obligatory requirement which is expected here.

============
Header files
============

oneDAL public identifiers are represented in the following header files:

.. list-table::
   :widths:  10 50
   :header-rows: 1

   * - header file

>      - description

Reviewer: Use capital letters (same for the other table)

>    * - ``onedal/onedal.hpp```

Reviewer: extra ` in the end

>      - Main header file of oneDAL library.

Reviewer: ```suggestion
     - The main header file of oneDAL library.
```

do you need say what it contains?

Author: Originally it was something like "including all public headers of oneDAL library" but it was decided to make it shorter.

   * - ``onedal/algo/%ALGORITHM%/%ALGORITM%.hpp``

>      - An algorithm header file for a particular algorithm.

Reviewer: I'd suggest adding a blank line between lines 16 and 17 to have two paragraphs here

>      - An algorithm header file for a particular algorithm.

Reviewer: ```suggestion
     - A header file for a particular algorithm.
```

>        The string %ALGORITHM% should be substituted with the name of the algorithm, for example, ``kmeans`` or ``knn``.

Reviewer: should %ALGORITHM%` be with rst markup?
```
``%ALGORITHM%``
```
same for utility, functionality, parent

   * - ``onedal/algo/misc/%FUNCTIONALITY%/%FUNCTIONALITY%.hpp``

>      -  A header file for miscellaneous data types and functionality intended to be used by oneDAL algorithms and applications for algorithm customization and optimization on various stages

Reviewer: this sentence is too long and hard-to-read, but I don't see how to break it in two. what about leaving only this information?

```suggestion
     -  A header file for miscellaneous data types and functionality that is intended to be used by oneDAL algorithms and applications.
```

>         of the analytical pipeline.

Reviewer: add a blank line after this one

        The string %FUNCTIONALITY% should be substituted with the functionality name, for example, ``mt19937`` or ``cross_enthropy_loss``.

   * - ``onedal/util/%UTILITY%.hpp```

>      - A header file for auxiliary functionality intended to be used for design of classes and implementation of methods such as memory allocators or type traits. 

Reviewer: a blank line after this one

>      - A header file for auxiliary functionality intended to be used for design of classes and implementation of methods such as memory allocators or type traits. 

Reviewer: ```suggestion
     - A header file for auxiliary functionality that is intended to be used to design classes and implement such methods as memory allocators and type traits. 
```

Author: The expected meaning is: things like memory allocators and type traits form auxiliary functionality which is used to design  ...

Reviewer: would this work?

```suggestion
     - A header file for auxiliary functionality intended to be used to design classes that implement such methods as memory allocators and type traits. 
```

>        The string %UTILITY% should be substituted with the auxiliary functionality name, for example, ``usm_allocator`` or ``type_traits``. 

Reviewer: q: do you need to have ``onedal/`` part for each header file? Why not mention before the table that all of these files are inside the ``ondeal`` directory?

concern: for the third row the path to the headerfile is too long, that causes the second column to be really small (sphinx does not break the inline code into multiple lines)
![image](https://user-images.githubusercontent.com/17865025/83018294-b6426b00-a02d-11ea-96a4-38fbe9f7b36c.png)

Reviewer: the solution I can think of is to break the path into the directory and the name of the file, and have something like this:
```
filename: ``%ALGORITM%.hpp``

located in: ``onedal/algo/%ALGORITHM%/``
```

not sure it's a good idea

Author: From my point of view, it is better to mention "onedal" folder before the table and omit it inside the table. Option with filename and location is also ok for me

     - Description
   * - ``onedal/onedal.hpp``
     - The main header file of oneDAL library.
   * - ``onedal/algo/%ALGO%/%ALGO%.hpp``
     - A header file for a particular algorithm.

       The string %ALGO% should be substituted with the name of the algorithm, for example, ``kmeans`` or ``knn``.

   * - ``onedal/algo/misc/%FUNC%/%FUNC%.hpp``
     -  A header file for miscellaneous data types and functionality that is intended to be used by oneDAL algorithms and applications.
        of the analytical pipeline.

>         The string %FUNC% should be substituted with the functionality name, for example, ``mt19937`` or ``cross_enthropy_loss``.

Reviewer: i think there is an extra space at the beginning of each of these lines
